lib/lambda_ethereum_consensus/store/state_store.ex

```diff
@@ -6,61 +6,80 @@ defmodule LambdaEthereumConsensus.Store.StateStore do
           {:ok, BeaconState.t()} | {:error, String.t()} | :not_found
-  def get_state(root) do
-    with {:ok, bin} <- root |> state_key() |> Db.get() do
+  def get_state(root, root_type) do
+    with {:ok, bin} <- root |> state_key(root_type) |> Db.get() do
       Ssz.from_ssz(bin, BeaconState)
     end
   end
 
-  @spec get_latest_state() ::
+  @spec get_latest_state(:block_root | :state_root) ::
           {:ok, BeaconState.t()} | {:error, String.t()} | :not_found
-  def get_latest_state do
-    last_key = root_by_slot_key(0xFFFFFFFFFFFFFFFF)
+  def get_latest_state(root_type) do
+    last_key = root_by_slot_key(0xFFFFFFFFFFFFFFFF, root_type)
 
     with {:ok, it} <- Db.iterate(),
          {:ok, _key, _value} <- Exleveldb.iterator_move(it, last_key),
-         {:ok, @stateslot_prefix <> _slot, root} <- Exleveldb.iterator_move(it, :prev),
+         {:ok, _key, root} <- Exleveldb.iterator_move(it, :prev) |> validate_key_type(root_type),
          :ok <- Exleveldb.iterator_close(it) do
-      get_state(root)
+      get_state(root, root_type)
     else
       {:ok, _key, _value} -> :not_found
       {:error, :invalid_iterator} -> :not_found
     end
   end
 
-  @spec get_state_root_by_slot(Types.slot()) ::
+  defp validate_key_type({:ok, key, root}, root_type) do
+    valid_prefix =
+      if root_type == :block_root,
+        do: @stateslot_prefix_by_block,
+        else: @stateslot_prefix_by_state
+
+    <<prefix::binary-size(byte_size(valid_prefix)), _::binary>> = key
+    if valid_prefix == prefix, do: {:ok, key, root}, else: {:error, :invalid_iterator}
+  end
+
+  @spec get_state_root_by_slot(Types.slot(), :block_root | :state_root) ::
           {:ok, Types.root()} | {:error, String.t()} | :not_found
-  def get_state_root_by_slot(slot),
-    do: slot |> root_by_slot_key() |> Db.get()
+  def get_state_root_by_slot(slot, root_type),
+    do: slot |> root_by_slot_key(root_type) |> Db.get()
 
-  @spec get_state_by_slot(Types.slot()) ::
+  @spec get_state_by_slot(Types.slot(), :block_root | :state_root) ::
           {:ok, BeaconState.t()} | {:error, String.t()} | :not_found
-  def get_state_by_slot(slot) do
+  def get_state_by_slot(slot, root_type) do
     # WARN: this will return the latest state received for the given slot
-    with {:ok, root} <- get_state_root_by_slot(slot) do
-      get_state(root)
+    with {:ok, root} <- get_state_root_by_slot(slot, root_type) do
+      get_state(root, root_type)
     end
   end
 
-  defp state_key(root), do: Utils.get_key(@state_prefix, root)
-  defp root_by_slot_key(slot), do: Utils.get_key(@stateslot_prefix, slot)
+  defp state_key(root, :block_root), do: Utils.get_key(@state_prefix_by_block, root)
+  defp state_key(root, :state_root), do: Utils.get_key(@state_prefix_by_state, root)
+  defp root_by_slot_key(slot, :block_root), do: Utils.get_key(@stateslot_prefix_by_block, slot)
+  defp root_by_slot_key(slot, :state_root), do: Utils.get_key(@stateslot_prefix_by_state, slot)
```

Comment:
It'd be simpler to just have four different functions.

Reply:
it will led to case .. do's in the other functions body but if you think so I can change it initially I have done it that way but things got messy